### PR TITLE
Order by composite fields

### PR DESCRIPTION
**Pull request by dpetrick**
### Overview
Expands `orderBy` capabilities to include ordering by composite fields.

- Over to-one composites:
```graphql
query { findManyTestModel(orderBy: { to_one_composite: { field: asc } }) { id } }
```

- Over a to-many composite aggregation (count):
```graphql
query { findManyTestModel(orderBy: { to_many_composite: { _count: asc } }) { id } }
```

- Over to-one relation: Order by to-one or to-many composite (aggregation):
```graphql
query { findManyTestModel(orderBy: { to_one_rel: { to_one_composite: { field: asc } } }) { id } }
```

- Ordering a to-many relation by their composite fields:
```graphql
query {
    findManyTestModel {
        id
        to_many_relation(orderBy: { to_one_composite: { field: asc } }) {
            id
            to_one_composite {
                field
            }
        }
    }
}
```

### What is not included
- Order by to-many composite selection:
```graphql
query {
  findManyModels {
  	to_many_composites(orderBy: { field: asc }) {
      field
    }
  }
}
```

- Order by scalar aggregation for `groupBy`.
```graphql
query {
  groupByModel(by: [field], orderBy: { to_many_composite: { field: asc} }) {
    _sum {
      field
    }
  }
}

```

### CODE REVIEW

File: query-engine/connectors/mongodb-query-connector/src/orderby.rs [hunk 1]
```diff
@@ -210,21 +261,22 @@ impl OrderByBuilder {
...
             // See `unwind_aggregate_joins` for more explanation
-            if let OrderBy::Aggregation(order_by_aggregate) = &data.order_by {
+            if let OrderBy::ToManyAggregation(order_by_aggregate) = &data.order_by {
                 if !order_by_aggregate.path.is_empty() {
                     match order_by_aggregate.sort_aggregation {
                         prisma_models::SortAggregation::Count => {
-                            if data.order_by.has_middle_to_one_path() {
+                            if let Some(clone_to) = data.prefix.as_ref().and_then(|x| x.clone_to.clone()) {
+                                order_aggregate_proj_doc.push(doc! { "$addFields": { clone_to.clone(): { "$size": { "$ifNull": [format!("${}", data.full_reference_path(false)), []] } } } });
+                                field_name = clone_to; // Todo: Just a hack right now, this whole function needs love.
```
Comment (garrensmith) [inline query-engine/connectors/mongodb-query-connector/src/orderby.rs:270]: Do we have a ticket for clean up work we want to do once we have launched mongo and have a little more time for clean up?

File: query-engine/connectors/sql-query-connector/src/ordering.rs [hunk 2]
```diff
@@ -125,23 +125,26 @@ pub fn compute_joins_aggregation(
...
 
     let aggregation_type = match order_by.sort_aggregation {
         SortAggregation::Count => AggregationType::Count,
         _ => unreachable!("Order by relation aggregation other than count are not supported"),
     };
+
     // We perform the aggregation on the last join
     let last_aggr_join = compute_aggr_join(
-        last_hop,
+        last_hop.into_relation_hop().unwrap(),
```
Comment (garrensmith) [inline query-engine/connectors/sql-query-connector/src/ordering.rs:141]: I'm assuming this `unwrap` and the next one are safe?

**Review by garrensmith — APPROVED** [review]
This is great work. I can't give a really deep review but the general code and test coverage looks good. Should we keep it in this branch until Pierre has done the dmmf work for it. The work can sometimes surface some bugs which we could then fix here.

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/queries/order_and_pagination/order_by_composite.rs [hunk 3]
```diff
@@ -0,0 +1,827 @@
...
+        Ok(())
+    }
+}
+
+#[test_suite(schema(to_many_composites), only(MongoDb))]
+mod to_many {
+    // Order a composite selection by a single orderBy.
+    // todo: Doesn't work yet at all, needs more in-depth code changes.
+    //#[connector_test]
+    // async fn simple_composite_selection_ordering(runner: Runner) -> TestResult<()> {
```
[the remark this answers is not in the document]
Comment (dpetrick) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/queries/order_and_pagination/order_by_composite.rs:202]: It's likely okay to not have this for GA, but I'm doing follow-up tickets anyways for Matt to decide.

File: query-engine/connectors/mongodb-query-connector/src/orderby.rs (hunk 1 above)
Comment (dpetrick) [reply query-engine/connectors/mongodb-query-connector/src/orderby.rs:270]: Let's just track tech debt tickets in the engine repo directly with "tech debt" label attached: https://github.com/prisma/prisma-engines/issues/2716

File: query-engine/connectors/sql-query-connector/src/ordering.rs (hunk 2 above)
Comment (dpetrick) [reply query-engine/connectors/sql-query-connector/src/ordering.rs:141]: Added a comment.

**Review by Weakky — COMMENTED** [review]
Did my best, couldn't spot anything fishy. I find tests slightly hard to reason about since they only fetch ids but the data is all the way at the bottom of the file so you constantly have to scroll up and down and remember which id has what data. It was too hard on a phone at least.
If relevant, I'd suggest either querying the relevant data so that it's visible on the snapshot or adding comments on the data that's being fetched (at the cost of having to update those comments if we ever change the data set).
I read the PR three times but I'd still wouldn't count on my approval as any strong guarantee given the circumstances.

File: query-engine/core/src/schema_builder/input_types/fields/arguments.rs [hunk 4]
```diff
@@ -136,21 +153,20 @@ pub(crate) fn many_records_arguments(
     args
 }
 
+/// Builds "many composite where" arguments for to-many composite selection sets.
+pub(crate) fn composite_selection_arguments(_ctx: &mut BuilderContext, _cf: &CompositeFieldRef) -> Vec<InputField> {
```
[the remark this answers is not in the document]
Comment (dpetrick) [reply query-engine/core/src/schema_builder/input_types/fields/arguments.rs:157]: Yea, this is a placeholder for the selection arguments down the road.

**Pull request merged**